File: ScalaEngine/src/main/scala/Orchestrator.scala
import java.util.UUID

import Orchestrator._
import akka.actor.{Actor, ActorLogging, Props}
import org.locationtech.spatial4j.context.SpatialContextFactory
import position.Position

import scala.util.Random

/**
  * Created by Ron on 19/06/2017.
  */
class Orchestrator extends Actor with ActorLogging {
import Orchestrator.{initSpeed,intiHeading}

  override def receive: PartialFunction[Any, Unit] ={
    case AddAircraft =>
      val uuid = UUID.randomUUID()
      context.actorOf(Aircraft.props(uuid,initSpeed,intiHeading,initPosition), uuid.toString)
      log.info("{}: created aircraft with uuid: {}",self.path.name,uuid)
    case _  => log.info("{}: received unknown message",self.path.name)
  }
}

object Orchestrator{
  case object AddAircraft
  val random = Random
  val MIN_SPEED = 9000;//meter/second
  val MAX_SPEED = 13000;//meter/second
  val MAX_LONGITUDE = 180.0
  val MAX_LATITUDE = 90.0
  val MIN_LONGITUDE: Double = -180.0
  val MIN_LATITUDE: Double = -90.0

  def initSpeed: Double = MIN_SPEED + random.nextDouble * MAX_SPEED
  def intiHeading: Double = random.nextDouble * 360
  def initPosition = new Position((MAX_LATITUDE - MIN_LATITUDE) * random.nextDouble() + MIN_LATITUDE,(MAX_LONGITUDE - MIN_LONGITUDE) * random.nextDouble() + MIN_LONGITUDE)
}

File: Engine/src/main/java/akkaLabs/ExtremeAircrafts/Orchestrator.java
package akkaLabs.ExtremeAircrafts;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.event.japi.LookupEventBus;
import akkaLabs.ExtremeAircrafts.commands.aircraft.ChangePosition;
import akkaLabs.ExtremeAircrafts.commands.aircraft.ModifyAircrafts;
import akkaLabs.ExtremeAircrafts.eventbus.PositionChangedEvelope;
import akkaLabs.ExtremeAircrafts.position.Position;
import com.google.inject.name.Named;
import org.locationtech.spatial4j.context.SpatialContext;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.UUID;
import java.util.stream.IntStream;

import static akkaLabs.ExtremeAircrafts.ExtremeModule.*;

public class Orchestrator extends AbstractActor {

	private final LoggingAdapter logger = Logging.getLogger(getContext().getSystem(), this);
	private final Map<UUID, ActorRef> uuidToActor;
	private final SpatialContext spatialContext;
	private int aircraftsCount;
	@Named(value="positionChangedEventBus")
	private LookupEventBus<PositionChangedEvelope, ActorRef, String> eventBus;

	public Orchestrator(SpatialContext spatialContext,LookupEventBus<PositionChangedEvelope, ActorRef, String> eventBus) {
		this.spatialContext = spatialContext;
		this.eventBus = eventBus;
		this.uuidToActor = new HashMap<>();
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder().
				match(ModifyAircrafts.class, msg -> {
					int n = msg.getNumOfAircrafts();
					if (this.aircraftsCount < n)
					{// add aircrafts
						IntStream.range(this.aircraftsCount, n).forEach(i ->
						{
							UUID uuid = UUID.randomUUID();
							logger.info("Creating actor #" + i + " uuid:" + uuid.toString());
							ActorRef newAircraft = getContext().actorOf(getProps(uuid), uuid.toString());

							double latitude = (MAX_LATITUDE - MIN_LATITUDE) * Math.random() + MIN_LATITUDE;
							double longitude = (MAX_LONGITUDE - MIN_LONGITUDE) * Math.random() + MIN_LONGITUDE;
							newAircraft.tell(new ChangePosition(new Position(longitude, latitude, 15)), ActorRef.noSender());
							uuidToActor.put(uuid, newAircraft);

							getContext().watch(newAircraft);
						});
					}
					else if (this.aircraftsCount > n) { // remove aircrafts
						Iterator<UUID> uuidIterator = uuidToActor.keySet().iterator();
						IntStream.range(n, this.aircraftsCount).forEach(i -> {
							UUID uuid = uuidIterator.next();
							logger.info("Stopping actor with uuid:" + uuid.toString());
							getContext().stop(uuidToActor.get(uuid));
						});
					}
					this.aircraftsCount = n;
				}).
				build();
	}

	public Props getProps(UUID uuid) {
		return Props.create(Aircraft.class, () -> new Aircraft(uuid, MIN_SPEED + Math.random() * MAX_SPEED, Math.random() * 360, this.spatialContext,this.eventBus));
	}

	// for tests
	public int getAircraftsCount()
	{
		return this.aircraftsCount;
	}
}